File: frontend/js/views/notifications.js
/* 通知中心视图 + 顶栏未读数轮询。 */
window.Views = window.Views || {};
window.Views.notifications = (function () {
  const KIND_LABEL = {
    task_created: "新任务",
    task_status: "状态变更",
    proposal_finalize: "方案 finalize",
    proposal_modified: "方案修改",
    review_added: "新回顾",
    ability_pending: "能力值待审",
    progress_update: "进度更新",
    info: "提示",
  };

  async function render(main) {
    const data = (await API.get("/api/notifications?limit=100")).data || [];
    main.innerHTML = "";
    const wrap = UI.el("div");
    wrap.appendChild(UI.el("div", { class: "page-header" }, [
      UI.el("h2", { class: "page-title" }, "通知中心"),
      UI.el("div", {}, [
        UI.el("button", { class: "btn", id: "ntf-mark-all" }, "全部已读"),
      ]),
    ]));

    if (!data.length) {
      wrap.appendChild(UI.el("div", { class: "empty" }, "暂无通知"));
      main.appendChild(wrap);
      return;
    }

    const list = UI.el("div");
    data.forEach((n) => list.appendChild(item(n, main)));
    wrap.appendChild(list);
    main.appendChild(wrap);

    document.getElementById("ntf-mark-all").onclick = async () => {
      await API.post("/api/notifications/mark_read", { ids: [], all: true });
      await refreshBadge();
      render(main);
    };
  }

  function item(n, main) {
    const div = UI.el("div", {
      class: "notif-item" + (n.read ? " read" : ""),
    });
    const kindLabel = KIND_LABEL[n.kind] || n.kind;
    div.innerHTML = `
      <div class="notif-row">
        <span class="badge">${UI.escape(kindLabel)}</span>
        <span class="notif-title">${UI.escape(n.title)}</span>
        <span class="muted notif-time">${UI.escape(n.created_at || "")}</span>
      </div>
      ${n.body ? `<div class="notif-body">${UI.escape(n.body)}</div>` : ""}
    `;
    const actions = UI.el("div", { class: "actions-row" });
    if (n.link) {
      const open = UI.el("a", { class: "btn", href: n.link }, "查看");
      open.onclick = async () => {
        if (!n.read) {
          await API.post("/api/notifications/mark_read", {
            ids: [n.id],
            all: false,
          });
          await refreshBadge();
        }
      };
      actions.appendChild(open);
    }
    if (!n.read) {
      const mk = UI.el("button", { class: "btn" }, "标记已读");
      mk.onclick = async () => {
        await API.post("/api/notifications/mark_read", {
          ids: [n.id],
          all: false,
        });
        await refreshBadge();
        render(main);
      };
      actions.appendChild(mk);
    }
    const del = UI.el("button", { class: "btn btn-danger" }, "删除");
    del.onclick = async () => {
      await API.del(`/api/notifications/${encodeURIComponent(n.id)}`);
      await refreshBadge();
      render(main);
    };
    actions.appendChild(del);
    div.appendChild(actions);
    return div;
  }

  async function refreshBadge() {
    try {
      const res = await API.get("/api/notifications/unread_count");
      const count = (res.data && res.data.count) || 0;
      const dot = document.getElementById("nav-notif-dot");
      if (!dot) return;
      if (count > 0) {
        dot.textContent = count > 99 ? "99+" : String(count);
        dot.classList.remove("hidden");
      } else {
        dot.classList.add("hidden");
      }
    } catch {
      /* 静默失败 */
    }
  }

  function startPolling() {
    refreshBadge();
    setInterval(refreshBadge, 30000);
  }

  return { render, refreshBadge, startPolling };
})();

File: backend/api/notifications.py
"""通知中心 API（UX §3.1 P2）。

事件源：方案 finalize / 提交回顾 / 能力值待审产生 / 任务状态变更 / 进度更新等。
存储：database.json 顶层 `notifications` 数组（按时间倒序），最多保留 200 条。
"""

from __future__ import annotations

from typing import Any, Literal, Optional

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from backend.core.storage import gen_id, get_db, now_iso
from backend.models.schemas import APIResponse

router = APIRouter(prefix="/api/notifications", tags=["notifications"])


MAX_KEEP = 200


class NotificationCreate(BaseModel):
    title: str
    body: Optional[str] = None
    kind: Literal[
        "task_created",
        "task_status",
        "proposal_finalize",
        "proposal_modified",
        "review_added",
        "ability_pending",
        "progress_update",
        "info",
    ] = "info"
    link: Optional[str] = None
    related: dict[str, Any] = Field(default_factory=dict)


class NotificationMark(BaseModel):
    ids: list[str] = Field(default_factory=list)
    all: bool = False


def push_notification(
    data: dict[str, Any],
    *,
    title: str,
    kind: str = "info",
    body: str | None = None,
    link: str | None = None,
    related: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """供其它后端模块直接调用：把通知写入事务中的 data 引用。"""

    notifications = data.setdefault("notifications", [])
    item = {
        "id": gen_id("ntf"),
        "title": title,
        "body": body or "",
        "kind": kind,
        "link": link,
        "related": related or {},
        "created_at": now_iso(),
        "read": False,
    }
    notifications.insert(0, item)
    if len(notifications) > MAX_KEEP:
        del notifications[MAX_KEEP:]
    return item


@router.get("")
def list_notifications(limit: int = 50, only_unread: bool = False) -> APIResponse:
    db = get_db()
    items = list(db.get_section("notifications") or [])
    if only_unread:
        items = [n for n in items if not n.get("read")]
    return APIResponse(ok=True, data=items[: max(1, min(limit, MAX_KEEP))])


@router.get("/unread_count")
def unread_count() -> APIResponse:
    db = get_db()
    items = db.get_section("notifications") or []
    n = sum(1 for x in items if not x.get("read"))
    return APIResponse(ok=True, data={"count": n})


@router.post("")
def create_notification(body: NotificationCreate) -> APIResponse:
    db = get_db()
    with db.transaction() as data:
        item = push_notification(
            data,
            title=body.title,
            body=body.body,
            kind=body.kind,
            link=body.link,
            related=body.related,
        )
    return APIResponse(ok=True, data=item)


@router.post("/mark_read")
def mark_read(body: NotificationMark) -> APIResponse:
    db = get_db()
    with db.transaction() as data:
        items = data.setdefault("notifications", [])
        changed = 0
        target = set(body.ids or [])
        for it in items:
            if body.all or it.get("id") in target:
                if not it.get("read"):
                    it["read"] = True
                    it["read_at"] = now_iso()
                    changed += 1
        return APIResponse(ok=True, data={"updated": changed})


@router.delete("/{nid}")
def delete_notification(nid: str) -> APIResponse:
    db = get_db()
    with db.transaction() as data:
        items = data.setdefault("notifications", [])
        new_items = [x for x in items if x.get("id") != nid]
        if len(new_items) == len(items):
            raise HTTPException(404, f"通知不存在: {nid}")
        data["notifications"] = new_items
        return APIResponse(ok=True, message="已删除")
